controller/scheduler-v2/scheduler.go

```diff
@@ -69,62 +80,91 @@ func (s *Scheduler) Run() error {
 	log.Info("starting scheduler loop")
 	defer log.Info("exiting scheduler loop")
 
-	jobTicker := time.Tick(30 * time.Second)
-	formationTicker := time.Tick(time.Minute)
-	formationTimeout := time.After(time.Second)
-	go func() {
-		for {
-			select {
-			case <-jobTicker:
-				triggerChan(s.jobSync)
-			case <-formationTicker:
-				triggerChan(s.formationSync)
-			case <-formationTimeout:
-				triggerChan(s.formationSync)
-			}
-		}
-	}()
-	triggerChan(s.jobSync)
+	tickChannel(s.syncJobs, 30*time.Second)
+	tickChannel(s.syncFormations, time.Minute)
+	tickChannel(s.syncHosts, 5*time.Minute)
+
+	stream, err := s.StreamHosts(s.hostChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream hosts. Error: %v", err)
+	}
+	defer stream.Close()
+
+	stream, err = s.StreamFormations(nil, s.formationChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream formations. Error: %v", err)
+	}
+	defer stream.Close()
+
+	go s.RunPutJobs()
 
 	for {
+		// First handle events reconciling our state with the cluster
 		select {
 		case <-s.stop:
 			return nil
-		case req := <-s.jobRequests:
-			s.jobRequestHandler(req)
-		case <-s.rectifyJobs:
-			s.RectifyJobs()
+		case h := <-s.hostChange:
+			s.followHost(h)
+			continue
+		case he := <-s.hostEvents:
+			s.handleHostEvent(he)
+			continue
 		case ef := <-s.formationChange:
-			s.formationChangeHandler(ef)
-		case <-s.formationSync:
+			s.FormationChange(ef)
+			continue
```

Comment:
We need to handle these channels closing.

We already do stream reconnects (with a [10s timeout](https://github.com/flynn/flynn/blob/master/pkg/httpclient/stream.go#L83-L86)), but I think we should also have goroutines in the scheduler reconnecting the streams in case things go down for a long time.